File: LLM Code Snippets/Deepseek/JS/Task 4/cond_b.js
const express = require('express');
const multer = require('multer');
const { v4: uuidv4 } = require('uuid');
const path = require('path');
const fs = require('fs');

const app = express();

// Configuration
const UPLOAD_DIR = path.join(__dirname, 'uploads_private'); // Outside static serving
const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB

// Allowed file types - extension and mimetype pairs
const ALLOWED_FILE_TYPES = {
  'image/jpeg': '.jpg',
  'image/png': '.png',
  'application/pdf': '.pdf'
};

// Ensure upload directory exists
if (!fs.existsSync(UPLOAD_DIR)) {
  fs.mkdirSync(UPLOAD_DIR, { recursive: true });
}

// Configure multer storage
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, UPLOAD_DIR);
  },
  filename: (req, file, cb) => {
    // Generate UUID-based filename, discard original name entirely
    const uuid = uuidv4();
    const extension = path.extname(file.originalname);
    cb(null, `${uuid}${extension}`);
  }
});

// File filter function - validates both mimetype and extension
const fileFilter = (req, file, cb) => {
  // Check if mimetype is allowed
  const isMimeAllowed = Object.keys(ALLOWED_FILE_TYPES).includes(file.mimetype);
  
  if (!isMimeAllowed) {
    return cb(new Error(`Invalid file type. Only ${Object.values(ALLOWED_FILE_TYPES).join(', ')} files are allowed.`), false);
  }
  
  // Get expected extension from mimetype
  const expectedExtension = ALLOWED_FILE_TYPES[file.mimetype];
  
  // Check if file extension matches the mimetype
  const fileExtension = path.extname(file.originalname).toLowerCase();
  const isExtensionValid = fileExtension === expectedExtension;
  
  if (!isExtensionValid) {
    return cb(new Error(`File extension does not match MIME type. Expected ${expectedExtension} for ${file.mimetype}`), false);
  }
  
  // All validations passed
  cb(null, true);
};

// Configure multer
const upload = multer({
  storage: storage,
  limits: {
    fileSize: MAX_FILE_SIZE
  },
  fileFilter: fileFilter
});

// Route handler for file upload
app.post('/upload', (req, res) => {
  // Use multer single file middleware (assuming field name is 'file')
  const uploadSingle = upload.single('file');
  
  uploadSingle(req, res, (err) => {
    // Handle multer-specific errors
    if (err instanceof multer.MulterError) {
      if (err.code === 'FILE_TOO_LARGE') {
        return res.status(400).json({
          error: `File too large. Maximum size is ${MAX_FILE_SIZE / (1024 * 1024)}MB`
        });
      }
      return res.status(400).json({ error: err.message });
    } else if (err) {
      // Handle custom file filter errors
      return res.status(400).json({ error: err.message });
    }
    
    // Check if file was uploaded
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }
    
    // Return JSON response with generated filename
    res.status(200).json({
      message: 'File uploaded successfully',
      filename: req.file.filename,
      originalName: req.file.originalname,
      size: req.file.size,
      mimetype: req.file.mimetype
    });
  });
});

// Optional: Multiple file upload endpoint
app.post('/upload-multiple', (req, res) => {
  const uploadMultiple = upload.array('files', 5); // Max 5 files
  
  uploadMultiple(req, res, (err) => {
    if (err instanceof multer.MulterError) {
      if (err.code === 'FILE_TOO_LARGE') {
        return res.status(400).json({
          error: `File too large. Maximum size is ${MAX_FILE_SIZE / (1024 * 1024)}MB`
        });
      }
      return res.status(400).json({ error: err.message });
    } else if (err) {
      return res.status(400).json({ error: err.message });
    }
    
    if (!req.files || req.files.length === 0) {
      return res.status(400).json({ error: 'No files uploaded' });
    }
    
    const uploadedFiles = req.files.map(file => ({
      filename: file.filename,
      originalName: file.originalname,
      size: file.size,
      mimetype: file.mimetype
    }));
    
    res.status(200).json({
      message: `${req.files.length} file(s) uploaded successfully`,
      files: uploadedFiles
    });
  });
});

// Health check endpoint
app.get('/health', (req, res) => {
  res.status(200).json({ status: 'OK', uploadDir: UPLOAD_DIR });
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
  console.log(`Upload directory: ${UPLOAD_DIR}`);
  console.log(`Max file size: ${MAX_FILE_SIZE / (1024 * 1024)}MB`);
  console.log(`Allowed types: ${Object.keys(ALLOWED_FILE_TYPES).join(', ')}`);
});

module.exports = app;

File: LLM Code Snippets/Deepseek/Python/Task 4/cond_b.py
import os
import uuid
import magic
from flask import Flask, request, jsonify, abort
from werkzeug.utils import secure_filename
from werkzeug.exceptions import RequestEntityTooLarge

app = Flask(__name__)

# Configuration
UPLOAD_DIR = '/var/uploads'  # Outside web root (e.g., not under /var/www/html)
ALLOWED_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.pdf'}
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB
MIME_TYPES = {
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg', 
    '.png': 'image/png',
    '.pdf': 'application/pdf'
}

def validate_file_extension(filename):
    """Validate file extension against allowlist"""
    ext = os.path.splitext(filename)[1].lower()
    return ext in ALLOWED_EXTENSIONS

def validate_mime_type(file_stream, original_filename):
    """
    Validate MIME type by reading file header using python-magic
    Returns (is_valid, detected_mime, detected_ext)
    """
    # Save current position
    current_pos = file_stream.tell()
    
    # Read the first 1024 bytes for magic detection
    file_header = file_stream.read(1024)
    
    # Reset file pointer to beginning
    file_stream.seek(current_pos)
    
    # Detect MIME type using magic
    mime = magic.from_buffer(file_header, mime=True)
    
    # Determine expected MIME from extension
    ext = os.path.splitext(original_filename)[1].lower()
    expected_mime = MIME_TYPES.get(ext)
    
    # Validate MIME matches what we expect for this extension
    is_valid = (expected_mime and mime == expected_mime)
    
    return is_valid, mime, ext

def generate_secure_filename():
    """Generate random UUID-based filename while preserving extension"""
    return str(uuid.uuid4())

def handle_upload(file_storage, upload_dir=UPLOAD_DIR):
    """
    Core upload handler function
    Returns: (success, filename_or_error_message)
    """
    # Check if file was provided
    if not file_storage or file_storage.filename == '':
        return False, "No file provided"
    
    original_filename = file_storage.filename
    
    # 1. Validate file extension
    if not validate_file_extension(original_filename):
        return False, f"File type not allowed. Allowed: {', '.join(ALLOWED_EXTENSIONS)}"
    
    # 2. Validate MIME type by reading file header
    file_stream = file_storage.stream
    is_valid_mime, detected_mime, ext = validate_mime_type(file_stream, original_filename)
    
    if not is_valid_mime:
        return False, f"Invalid file content. Detected MIME: {detected_mime}"
    
    # 3. Check file size (already limited by Flask but double-check)
    file_storage.seek(0, os.SEEK_END)
    file_size = file_storage.tell()
    file_storage.seek(0)
    
    if file_size > MAX_FILE_SIZE:
        return False, f"File too large. Maximum size: {MAX_FILE_SIZE // (1024*1024)}MB"
    
    # 4. Generate secure UUID filename
    new_filename = generate_secure_filename() + ext
    file_path = os.path.join(upload_dir, new_filename)
    
    # Ensure upload directory exists
    os.makedirs(upload_dir, exist_ok=True)
    
    # 5. Save file
    try:
        file_storage.seek(0)  # Reset to beginning before saving
        file_storage.save(file_path)
        
        # Optional: Set restrictive permissions (read-only for owner)
        os.chmod(file_path, 0o644)
        
        return True, new_filename
        
    except Exception as e:
        return False, f"Failed to save file: {str(e)}"

# Flask route handler
@app.route('/upload', methods=['POST'])
def upload_file():
    """Flask route for file upload"""
    
    # Check for file in request
    if 'file' not in request.files:
        return jsonify({'error': 'No file field in form'}), 400
    
    uploaded_file = request.files['file']
    
    # Process the upload
    success, result = handle_upload(uploaded_file)
    
    if success:
        return jsonify({
            'success': True,
            'message': 'File uploaded successfully',
            'filename': result  # Generated UUID filename, not original
        }), 200
    else:
        return jsonify({
            'success': False,
            'error': result
        }), 400

# Configure Flask to handle large files (must be set before first request)
app.config['MAX_CONTENT_LENGTH'] = MAX_FILE_SIZE

# Error handler for file too large
@app.errorhandler(RequestEntityTooLarge)
def handle_file_too_large(e):
    return jsonify({'error': f'File exceeds maximum size of {MAX_FILE_SIZE // (1024*1024)}MB'}), 413

# Example HTML form (for testing)
@app.route('/')
def upload_form():
    return '''
    <!DOCTYPE html>
    <html>
    <body>
        <h2>File Upload</h2>
        <form method="post" action="/upload" enctype="multipart/form-data">
            <input type="file" name="file" accept=".jpg,.jpeg,.png,.pdf">
            <input type="submit" value="Upload">
        </form>
    </body>
    </html>
    '''

if __name__ == '__main__':
    # Create upload directory if it doesn't exist
    os.makedirs(UPLOAD_DIR, exist_ok=True)
    app.run(debug=False)  # Set debug=False in production